### PR TITLE
WIP: Build Instructions

### PR DESCRIPTION
**Pull request by elsiehupp**
I'm marking this as "WIP" even though it should basically be ready to merge because I'd like some feedback before it gets approved, and I'm not sure where else to have the conversation other than here.

Thus far my changes mainly incorporate the macOS build instructions discussed in #2893, but I'd be happy to broaden the scope to Linux and possibly even Windows if I'm able to set up a test environment.

I also plan on updating the main [README.md](https://github.com/nextcloud/desktop/blob/0bb6dcebd878e2b2236006295fa291c6d23ae857/README.md) and the [Wiki](https://github.com/nextcloud/desktop/wiki) to reference these build instructions, so that could be added to the pull request.

A few initial questions: 

1. I removed the step for installing PdfLatex because the document is titled "Appendix A: Building the Client", and building the documentation seems like it would be separate from that.
   * Has the CMake script been updated so that there isn't a warning if the user builds the client without PdfLatex installed? My goal is that the instructions should have the user end up with a warning-free build.
   * If not in the main macOS client-building instructions, where *should* the documentation-building instructions go? Right now there is a [README.md](https://github.com/nextcloud/desktop/blob/250fcf380b682b646c8b30e35c17f6151fe31676/doc/README.md) that is not particularly discoverable.
   * Are there any other documentation-specific dependencies that can split off similar to PdfLatex?

2. I noticed that the introduction to the document includes a reference to the different Git URLs. Considering the cloning step in the macOS build instructions is quite wordy at this point, and the commentary on the URLs there is applicable to Linux and Windows, as well, should there maybe be a section consolidating the Git instructions, with internal URLs to scroll up or down to it? And considering Git instructions are applicable to the documentation, as well, should it perhaps go under its own top-level heading?

3. Despite the existing note to the contrary, the instructions don't actually put the build inside a `.pkg` installer. If that part is separate, should there be a mini-section devoted to it, like with the documentation? And what would be the command-line string for building a `.pkg` installer?

4. If I'm working on more than just the macOS build instructions, how should I do pull requests? Should I do a new one every couple commits on the same branch? If so, how should I solicit feedback without fragmenting the conversation thread?

Thank you in advance for your responses!

### CODE REVIEW

File: doc/building.rst
Comment (FlexW) [inline doc/building.rst:82]: I would do
```
export OPENSSL_ROOT_DIR=$(brew --prefix openssl)
QT_INSTALL_DIR=$(brew --prefix qt)
export CMAKE_PREFIX_PATH=${QT_INSTALL_DIR}/lib/cmake/Qt5:${QT_INSTALL_DIR}/lib/cmake/Qt5LinguistTools:${CMAKE_PREFIX_PATH}
```
That way the CMake finds the correct Qt5 CMake config files.
I also would not put them into `.zshrc` you may want to develop other applications on your machine and don't want your environment polluted by one project. My suggestion would be to replace `~/.zshrc` with for example `~/prepare_desktop_dev_env.sh`. Every time you want to build you then need to source this file with `source ~/prepare_desktop_build_env.sh`. If this is to much work for you, you can use a tool like [direnv](https://direnv.net/) and place the `prepare_desktop_build_env.sh` in the desktop repo.

File: doc/building.rst
Comment (FlexW) [inline doc/building.rst:67]: Are you sure that is really needed? I didn't had to do that on my vm (MacOS Catalina). I would add a comment to this that says if you get permission erros, than fix the permissions like this...

File: doc/building.rst
Comment (FlexW) [inline doc/building.rst:73]: `doxygen` is only needed for documentation (maybe you can add a note :)). I think `qt5` is also not needed, `qt` should be enough but maybe in the future it's better to be explicit about `qt5`, so maybe we should just keep `qt5`.

File: doc/building.rst
Comment (FlexW) [inline doc/building.rst:89]: I would drop the explanation about the appearance of the prompt. It's misleading. Everyone can change it and does so. If you want to find out if you are running bash or zsh you can check if the variable BASH_VERSION or ZSH_VERSION is set. If BASH_VERSION is set, you are running on bash. If the ZSH_VERSION variable is set, you running on zsh. But see my other comment what I think about `.zshrc` :)

File: doc/building.rst
Comment (FlexW) [inline doc/building.rst:137]: We don't use any submodules. This is not needed.

**Comment by elsiehupp** [comment]
One thought I have in terms of organizing the different sections is that "Appendix A: Building the Client" could be renamed "Appendix A: Contributing" (like the main headline in the root-directory README.md), and the two documents could be combined and directly (albeit manually) mirrored with each other. The hierarchy of subheadings is kind of another question entirely.

File: doc/building.rst
Comment (FlexW) [inline doc/building.rst:152]: I also needed to set `-D OPENSSL_CRYPTO_LIBRARY=$(brew --prefix openssl)/lib/libcrypto.dylib -D OPENSSL_SSL_LIBRARY=$(brew --prefix openssl)/lib/libssl.dylib`. This was necessary because it seems like that CMake does detect the system libraries first but we can not use them because they are not really OpenSSL libraries but LibreSSL libraries. They just happen to have the same names. From my understanding that is a problem in CMake's `FindOpenSSL.cmake`. But maybe someone knows it better than me;)

**Comment by elsiehupp** [comment]
FWIW GitHub does render `.rst` files in the same style that they render `.md` files, so just using the `.rst` files directly with GitHub is, in fact, a feasible option. The only thing is I don't know if GitHub would render `README.rst` in the same way that it renders `README.md`, though I could look into that.

For reference, this is what `building.rst` looks like in GitHub: https://github.com/nextcloud/desktop/blob/cace0dbc9a88afa69ec02e4b221077d62ed06939/doc/building.rst

**Review by FlexW — CHANGES_REQUESTED** [review]

File: doc/building.rst
Comment (elsiehupp) [reply doc/building.rst:82]: I did this initially, but the problem with putting `$(brew --prefix <package>)` in your shell profile is that it runs every time you open a new shell, which can slow things down substantially. I don't know the differences between the various shell-profile files, though, so the loading time may be less of an issue if it can be made to be invoked less frequently.

FWIW you do know that I suggested just putting the calls to `brew --prefix` in the CMake file (which would reduce how often they run), but Kevin shot that down.

Comment (FlexW) [reply doc/building.rst:82]: > I did this initially, but the problem with putting $(brew --prefix <package>) in your shell profile is that it runs every time you open a new shell, which can slow things down substantially. I don't know the differences between the various shell-profile files, though, so the loading time may be less of an issue if it can be made to be invoked less frequently.

Yes that's one reason why I suggest that you don't put the statements into `.zshrc`. Then they will not get executed if not needed. If you use absolute paths you may run into problems if the path changes. With using brew prefix the instructions are more future proof. 

> FWIW you do know that I suggested just putting the calls to brew --prefix in the CMake file (which would reduce how often they run), but Kevin shot that down.

Yes, because that would make brew a requirement of an macos build but you don't need to use brew to build the client. You could also install the software by hand and maybe later we start to use a package manager like [Conan](https://conan.io/) to ship our build dependencies and then the CMake files would not work anymore. CMake is just a (meta) build system, not a package (dependency) manager. CMake's only task is it to create the correct build files. Managing dependencies is out of the scope of CMake :)

**Comment by FlexW** [comment]
> Thus far my changes mainly incorporate the macOS build instructions discussed in #2893, but I'd be happy to broaden the scope to Linux and possibly even Windows if I'm able to set up a test environment.

I recently went through setting up my build environment on all platforms and different Linux. I could provide you with the build instructions for Windows and Linux that I used and noted down :) They just need some explaining words around them.

File: doc/building.rst
Comment (elsiehupp) [reply doc/building.rst:82]: My thought with these instructions is that it's more reasonable, for example, to explicitly tell people to use Homebrew than to bury that in the CMake. And the reason I put the paths in `.zshrc` is that it's tedious to have to reenter (i.e. copy and paste) the same information every time you run `cmake`. One option would be to tell the user (a) put it in a shell script, or (b) set it up in such-and-such a way on VS Code, but then that would assume the user is setting up VS Code or is otherwise going to set up a shell script.

How would you recommend I streamline the instructions so that the user doesn't have to run more than `cmake` and `make` every time they want to test an incrementally updated build (and assuming there could be a new shell session)? Again, the question is how you'd recommend **streamlining** the instructions.

File: doc/building.rst
Comment (elsiehupp) [reply doc/building.rst:67]: This is apparently a known issue under some circumstances. It may no longer be applicable, but I ran into it rather more recently. If you think I should do a troubleshooting section, then that's fine.

File: doc/building.rst
Comment (elsiehupp) [reply doc/building.rst:73]: I don't remember why I specified `qt5`. I think it's a holdover from the previous instructions. The current Homebrew formula for `qt5` apparently just an alternate name for the formular for `qt`: https://formulae.brew.sh/formula/qt

File: doc/building.rst
Comment (elsiehupp) [reply doc/building.rst:89]: The important thing is if the user needs to know what shell they're using, the instructions should specify a way of finding out. While in your personal circle it may be the norm for people to change the shell prompt, the sort of person who doesn't know what shell they're using presumably hasn't.

File: doc/building.rst
Comment (elsiehupp) [reply doc/building.rst:152]: If you set the `.zshrc` the way I specified in this version of instructions, this is not a problem. If you specify the variables some other way, it might very well be. I tested the instructions in #2983 multiple times. I can change this if it becomes necessary with other changes to the instructions.

**Comment by elsiehupp** [comment]
> > Thus far my changes mainly incorporate the macOS build instructions discussed in #2893, but I'd be happy to broaden the scope to Linux and possibly even Windows if I'm able to set up a test environment.
> 
> I recently went through setting up my build environment on all platforms. I could provide you with the build instructions for Windows and Linux that I used and noted down :) They just need some explaining words around them.
That would be helpful! You can drop them here or link to them in a gist!

File: doc/building.rst
Comment (elsiehupp) [reply doc/building.rst:82]: I should also clarify that I intend to modify the disclaimer at the top of the instructions and add a request that people open a GitHub Issue if the instructions don't work. This way, if the absolute path changes with future versions of Homebrew or whatever, the instructions can be updated to reflect that.

Comment (FlexW) [reply doc/building.rst:82]: Normally you run ` cmake` once when you clone the project. After that, you don't need to run it anymore unless you want to change the build configuration. With my proposal you would need to run `source ~/prepare_desktop_dev_env.sh` once a day if you open your terminal or if you open a new terminal in which you want to build. As I said if this is too much work for you, you can use a tool like direnv (I don't know if it's available for mac but anyway I would say that is out of the scope for the build instructions, that's a personal decision). 
Another idea: If your goal is to run the least possible commands, you can write your own shell script that exports the correct variables if they are not set and then runs cmake and make (that's what I do and it's really not difficult to write such a script. You could even start the app directly from that script. If you are really motivated, you could put such a script at the end of the build instructions;) ).

I would even say that it is harmful to put the instructions into `.zshrc` because you may get errors or side effects that you don't expect.

I think of the build instructions more like what steps need to be taken to build the app and not how your workflow should be because as you said people use different editors or IDEs. (If I were to write the build instructions I would not even put the export statements into a file, I would give the user a hint that she may want to put these instructions into a file to make her life easier) :)

File: doc/building.rst
Comment (FlexW) [reply doc/building.rst:89]: The user just needs to enter `echo $BASH_VERSION` and press enter. If this gives an output, you are using bash. If not, enter `echo $ZSH_VERSION`. If this gives you an output you are running zsh. If both commands give no output you are most likely running a different shell. I think this method is more reliable because one day the developers of zsh change the default prompt and then the instructions become out of date (the developers of zsh could also decide to drop ZSH_VERSION but I would consider this more unlikely).

File: doc/building.rst
Comment (FlexW) [reply doc/building.rst:152]: Interestingly I rerun my build without specifying them and there were no errors. Maybe I had a typo the last time. I guess we then don't need to add that.

File: doc/building.rst
Comment (elsiehupp) [reply doc/building.rst:82]: I just checked, and the following does in fact work:
```
% echo 'some string' >> ~/non_existent_file
```
Which is to say I could safely replace `~/.zshrc` with something like `~/.nextcloud_build_variables` (no extension needed, since it's not executable) and export them with `source ~/.nextcloud_build_variables` or `. ~/.nextcloud_build_variables` in a later step (i.e. right before running `cmake ..`). This would allow the environment to use up-to-the-minute `brew --prefix` paths while avoiding the necessity to run `brew --prefix` multiple times with every new shell session.

Of course, alternately, yes, the option for a build script could work, too.

> (If I were to write the build instructions I would not even put the export statements into a file, I would give the user a hint that she may want to put these instructions into a file to make her life easier)

I plan on having the build-instructions disclaimer explain that the instructions are not the only way of doing things; they are just a consistent and straightforward way of explaining how to do things.

File: doc/building.rst
Comment (elsiehupp) [reply doc/building.rst:89]: Of course, moving the environmental variables into a non-shell-specific source file eliminates the need to make a distinction between `bash` and `zsh` in the first.

**Review by elsiehupp — COMMENTED** [review]
@FlexW how does this version of the instructions look to you?

File: doc/building.rst
Comment (elsiehupp) [reply doc/building.rst]: How does [this](https://github.com/nextcloud/desktop/pull/2931/commits/801d20e3b0c966132943b6a139df2871aedaf56d) look?

**Comment by elsiehupp** [comment]
One problem with removing the documentation-only packages from the client build instructions is that `cmake` will still throw warnings if it can't find them. Under what circumstances does `cmake` actually build the documentation? The main relevant portion of `CMakeLists.txt` seems to be here:

https://github.com/nextcloud/desktop/blob/642320c007e02c3e1921a0060c15797598039415/CMakeLists.txt#L240-L250

Basically, if the build instructions don't include the dependencies for building the documentation, then the build instructions should explicitly opt out of building the documentation and explicitly opt out of calling the dependencies that haven't necessarily been installed.

Right now, the relevant conditional seems to be `if(NOT BUILD_LIBRARIES_ONLY)`, but the option `BUILD_LIBRARIES_ONLY` doesn't seem applicable to this particular scenario:

https://github.com/nextcloud/desktop/blob/642320c007e02c3e1921a0060c15797598039415/CMakeLists.txt#L165-L166

One option could be to create another build option, `BUILD_CLIENT_ONLY`, and use it in conditionals around the calls to Sphinx, PdfLatex, and Doxygen, as well as in the above conditional that adds the `doc` folder, but @er-vin seems very protective of the CMake scripts...

Again, the issue here is that a stated goal of the instructions I'm writing is that they need to not throw any build warnings or errors. Where there *are* build warnings or errors, either the instructions need to be changed or the underlying cause of the warnings or errors need to be fixed or suppressed. Which do you think is better? (And, no, I'm not okay with doing away with the goal of the instructions not throwing any build warnings or errors.)

File: doc/building.rst
Comment (FlexW) [inline doc/building.rst:141]: You seem to insist to not use `CMAKE_PREFIX_PATH`. Any reasons? :) But if this way works too it's also okay:)

File: doc/building.rst
Comment (FlexW) [inline doc/building.rst:176]: Not wrong but unnecessary if you have taken step `7` already. Could then be simplified to `cd build`.

File: doc/building.rst
Comment (FlexW) [inline doc/building.rst:178]: I would not use `~/Builds`. I would be more explicit and call it `~/desktop-install` or something similar because you don't want other builds then the desktop build in this directory.

**Comment by FlexW** [comment]
> @FlexW how does this version of the instructions look to you?
> 
> EDIT: I don't know if this comment attached in the right place, but it's in reference to the environmental variables.

It looks great :) I would say we can merge that.

Regarding the issues with building the documentation: I think the best would be to have an extra option to enable building the documentation and if it is enabled, doxygen etc. become required packages.

File: doc/building.rst
Comment (camilasan) [reply doc/building.rst]: it does look like a markdown file.

File: doc/building.rst
Comment (camilasan) [reply doc/building.rst]: Could we change the text to what @FlexW suggested?

File: doc/building.rst
Comment (camilasan) [reply doc/building.rst:176]: It is better to be detailed.

File: doc/building.rst
Comment (camilasan) [reply doc/building.rst]: We used to have it.

File: doc/building.rst
Comment (camilasan) [reply doc/building.rst]: > export them with `source ~/.nextcloud_build_variables` or `. ~/.nextcloud_build_variables` in a later step (i.e. right before running `cmake ..`). This would allow the environment to use up-to-the-minute `brew --prefix` paths while avoiding the necessity to run `brew --prefix` multiple times with every new shell session.

I think this text needs to be used in this section instead of just "The file ``~/.nextcloud_build_variables`` is just a suggestion for  convenience. You can use a different file or create an entire shell  script, but this way of doing things is the simplest to explain."

**Comment by camilasan** [comment]
>   * If not in the main macOS client-building instructions, where _should_ the documentation-building instructions go? Right now there is a [README.md](https://github.com/nextcloud/desktop/blob/250fcf380b682b646c8b30e35c17f6151fe31676/doc/README.md) that is not particularly discoverable.

What do you mean with not being discoverable? That is the table of content you can see here
https://docs.nextcloud.com/desktop/3.1/index.html#nextcloud-desktop-client-manual

>   * Are there any other documentation-specific dependencies that can split off similar to PdfLatex?

I think so, qt5keychain needs to be built for Windows for instance.

> 
> * I noticed that the introduction to the document includes a reference to the different Git URLs. Considering the cloning step in the macOS build instructions is quite wordy at this point, and the commentary on the URLs there is applicable to Linux and Windows, as well, should there maybe be a section consolidating the Git instructions, with internal URLs to scroll up or down to it? And considering Git instructions are applicable to the documentation, as well, should it perhaps go under its own top-level heading?

That sounds good.

> * Despite the existing note to the contrary, the instructions don't actually put the build inside a `.pkg` installer. If that part is separate, should there be a mini-section devoted to it, like with the documentation? And what would be the command-line string for building a `.pkg` installer?

That should be a separate part indeed.

> * If I'm working on more than just the macOS build instructions, how should I do pull requests? Should I do a new one every couple commits on the same branch? If so, how should I solicit feedback without fragmenting the conversation thread?

I would do one PR for each.

**Comment by camilasan** [comment]
Thanks a lot @elsiehupp for these changes, this is super valuable :heart: and I am sorry for the nitpicking :)

File: doc/building.rst
Comment (elsiehupp) [reply doc/building.rst:178]: On Linux or Windows I might agree, but on macOS the entire application is encapsulated in a folder with a `.app` extension, so it's safe to put it in a more generic "builds" folder. I haven't built Nextcloud Desktop on Linux or Windows, yet, though, so it might turn out that the `make` scripts create a subfolder on one or both of those, too.

File: doc/building.rst
Comment (elsiehupp) [reply doc/building.rst:176]: I mentioned this in https://github.com/nextcloud/desktop/issues/2893, but I put the full path in the `cd` each time because I want it to be less of an issue if the user goes back and does the steps out of order.

File: doc/building.rst
Comment (elsiehupp) [reply doc/building.rst:141]: I might have done it wrong, but `CMAKE_PREFIX_PATH` didn't seem to work as an environmental variable (rather than a suffix to the `cmake` command). I haven't tried it again, though.

**Comment by elsiehupp** [comment]
@camilasan with regard to the less-discoverable documentation, I meant that there aren't really links between the docs.nextcloud.com "building" page and the various GitHub markdown files. In the case of the README.md in the `doc` folder, it doesn't seem to have any links leading to it, so the only way you'd find your way to it would be if you blindly clicked on the folder in the top-level directory. As I mentioned before, the fact that GitHub seems to render `rst` files similarly to how it renders `md` files could allow us to have some files serve double-duty, which would help avoid having redundant information get out of sync.

File: doc/building.rst
Comment (elsiehupp) [reply doc/building.rst:141]: I tried it again, and it worked.

File: doc/building.rst
Comment (elsiehupp) [inline doc/building.rst:134]: @camilasan how is this updated wording?

File: doc/building.rst
Comment (elsiehupp) [inline doc/building.rst:213]: @FlexW obviously this needs actual instructions... Do you know what the `make` command is for building a `.pkg` installer?

File: doc/building.rst
Comment (elsiehupp) [inline doc/building.rst:131]: @FlexW with the updated CMakeLists.txt, would there be any warnings if the documentation-specific dependencies are not present when building the client?

**Comment by elsiehupp** [comment]
Regarding files doing double-duty, it might be possible to use [pandoc](https://pandoc.org/) in the CI workflow to autogenerate, say, the main `README.md` from a future version of `building.rst`. This could be a workaround for the slight incompatibilities between the two formats, but I would like to experiment with GitHub's `rst` implementation a bit, first.

**Comment by elsiehupp** [comment]
If you take a look at [this experimental commit](https://github.com/elsiehupp/desktop/tree/92d50f4ad92ea9ca5e80a5a84bc87492946af263) where I replaced the top-level `README.md` with a symbolic link named `README.rst`, GitHub does actually render it below the directory listing. There is definitely some wonkiness that would need to be accommodated were the existing `README.md` merged with `building.rst`, but it seems like it might actually be technically workable to have a single file serve both roles.

**Comment by elsiehupp** [comment]
Once we resolve the last few little things with the current draft, it should be ready to merge. The next steps after that might be incorporating the text of the GitHub Wiki into the documentation or incorporating Felix's Linux and Windows build instructions, or maybe both at the same time.

Considering the text on `building.rst` is getting a bit unwieldy in length, it might also be worthwhile splitting it into multiple pages, so I can look into what that might entail.

**Comment by elsiehupp** [comment]
By the way, @camilasan the way you can start your own mini-threads in a pull request is by clicking the "Files Changed" tab, selecting the scope under "Changes from...", then, when mousing over individual changed lines, clicking the "+" button to add your comments. Then, once you're done adding comments, you can submit your review with the green button in the upper right.
Screen Shot 2021-02-17 at 3 48 50 PM
(It took me a little while to figure all this out.)

**Comment by elsiehupp** [comment]
> If you take a look at [this experimental commit](https://github.com/elsiehupp/desktop/tree/92d50f4ad92ea9ca5e80a5a84bc87492946af263) where I replaced the top-level `README.md` with a symbolic link named `README.rst`, GitHub does actually render it below the directory listing. There is definitely some wonkiness that would need to be accommodated were the existing `README.md` merged with `building.rst`, but it seems like it might actually be technically workable to have a single file serve both roles.

Another option could be a Markdown Sphinx parser, like [`recommonmark`](https://github.com/readthedocs/recommonmark). I don't know what would be involved in some of the other aspects of building the documentation (such as internal links), but something like `recommonmark` could allow at the very least the GitHub-based developer documentation to be directly incorporated into the published documentation.

**Comment by elsiehupp** [comment]
One other thing: having `CMAKE_BUILD_PREFIX` default to `/usr/local` is probably not a good idea on macOS, if only because it requires `sudo make install` rather than just `make install` (though also `/usr` is a hidden folder, so it's inconvenient to navigate to it in the Finder). It could be a sensible default for macOS for `CMAKE_BUILD_PREFIX` to be `~/Builds` or, barring that, just `~` (or even `~/Desktop`).

**Comment by FlexW** [comment]
@elsiehupp [Here](https://cloud.nextcloud.com/s/6PQtijcCpLGGmZ8) you can find my notes for building the client on Windows, Linux (Ubuntu and Arch) and OSX. Please let me know if something is unclear.

File: doc/building.rst
Comment (FlexW) [reply doc/building.rst:131]: With the updated cmake scripts you get after the configure phase a summary that tells you which required and optional features are enabled. But anyway if the configure phase fails you get at the end an error. I think it should be clear from the output if there are any errors.

File: doc/building.rst
Comment (FlexW) [reply doc/building.rst:213]: This unfortunate not as easy as just invoking an make command. I have never done this myself by hand so I can't guide you here.

File: doc/building.rst
Comment (FlexW) [inline doc/building.rst:270]: Same here. I think it's better to first merge the build instructions and then later in a separate pr focus on the installer instructions.

**Comment by FlexW** [comment]
> One other thing: having `CMAKE_BUILD_PREFIX` default to `/usr/local` is probably not a good idea on macOS, if only because it requires `sudo make install` rather than just `make install` (though also `/usr` is a hidden folder, so it's inconvenient to navigate to it in the Finder). It could be a sensible default for macOS for `CMAKE_BUILD_PREFIX` to be `~/Builds` or, barring that, just `~` (or even `~/Desktop`).

I think you mean `CMAKE_INSTALL_PREFIX`? Yes, don't use `/usr/local` for development builds. You don't want to pollute your system with development builds.

File: doc/building.rst
Comment (elsiehupp) [reply doc/building.rst:131]: I can follow this up if necessary once the new CMake scripts get merged.

File: doc/building.rst
Comment (elsiehupp) [reply doc/building.rst]: Yes, we can revisit the installer build instructions later.

**Comment by elsiehupp** [comment]
> I think you mean `CMAKE_INSTALL_PREFIX`?

Yes, that's what I meant. I just typed it from memory rather than copying and pasting it... 😬

> Yes, don't use `/usr/local` for development builds. You don't want to pollute your system with development builds.

I can take a look at where the default build location is set in the CMake scripts and suggest a change there. Then once that's merged, I can remove the warning here.

**Comment by elsiehupp** [comment]
> @elsiehupp [Here](https://cloud.nextcloud.com/s/6PQtijcCpLGGmZ8) you can find my notes for building the client on Windows, Linux (Ubuntu and Arch) and OSX. Please let me know if something is unclear.

Thanks! This should be really helpful for updating the other build instructions!

**Comment by elsiehupp** [comment]
@FlexW I think aside from any last-minute feedback from @camilasan this should be ready to merge, if you want.

File: doc/building.rst
Comment (FlexW) [inline doc/building.rst:333]: Could you remove these two lines that update the submodules?  They are not necessary.

**Comment by FlexW** [comment]
Thank you @elsiehupp This looks really good :) Thanks for this contribution. This will help a lot of people to get started faster. Could you incorporate that last change I missed to tell you before? I know the Windows instructions need a update anyway. After that I will happily merge it:)

File: doc/building.rst
Comment (er-vin) [reply doc/building.rst:333]: I'm not so sure, we do have a submodule! It's use should be repaired though, it's the one which would reenable having proper crash reports in the wild.

Also the two submodules steps could be compressed as: `git submodule update --init`.

Comment (elsiehupp) [reply doc/building.rst]: I just replaced the two lines with the consolidated one. FWIW that's actually a section I hadn't really touched, so what you're commenting was already there before I started my work. I had actually already removed that bit from the main macOS instructions, though I could easily re-add it.

**Comment by elsiehupp** [comment]
Well, that was interesting!

Different-Types-of-Squash

**Comment by FlexW** [comment]
@elsiehupp Maybe I should have been clearer with the squashing. The idea was to just have a single commit because you changed just one thing and that are the build steps for OSX.

The thing is, a lot of the commits you made are not useful on their own and if you later look at the history of the project you see a lot of weird commits and ask yourself for what they are. Now, these useless commits are still in your commit message. 

Just give your commit message a meaningful description like: Update OSX build instructions (or something else I'm sure there are better messages). Just one other thing to keep in mind: The first line of the commit should not be longer than 50 characters. If you need to add an additional description to the commit leave one line blank and then write the description. The lines of the description should not have more than 80 characters.

See also [here](https://chris.beams.io/posts/git-commit/) how to write good commit messages.

**Comment by elsiehupp** [comment]
@FlexW thanks for the link on how to write good Git messages. FWIW the incremental Git messages are mostly useful as a work-in-progress thing (i.e. if I want to revert one small thing from a while back), but I definitely see it being clutter if those incremental commits are all shown in the main branch.

**Review by FlexW — APPROVED** [review]

**Pull request merged**
